numpy/_core/tests/test_ufunc.py

```diff
@@ -547,6 +547,9 @@ def test_partial_signature_mismatch_with_cache(self):
         with pytest.raises(TypeError):
             np.add(np.float16(1), np.uint64(2), sig=("e", "d", None))
 
+    @pytest.mark.xfail(np._get_promotion_state() != "legacy",
+            reason="NEP 50 impl breaks casting checks when `dtype=` is used "
```

Comment:
This is a tricky issue to solve in a generic way, because what we need is a way to define cast safety for scalars generically (additionally it is a bit).

It is on the tracking issue, in most cases I think it is very niche (`dtype=` use is very rare, and this only really matters for `dtype=int` with Python floats input).

You can put a clunky thing in place to make it work for Python ints explicitly, but not sure this isn't so niche that it is a priority?

Reply:
I agree it is super niche; I think the main use case might be where people rely on "random" input to be cast to something as part of an operation (say, `np.substract(a, b, dtype='f8')` when `a` and `b` are passed-in arguments.

Anyway, my comment here was mostly a question of whether if we don't fix it now (which is fine), it is worth raising a separate issue about it so we will fix it eventually